File: searchinRoratedSOrtedArray.java
    /*  Explanation
    Time Complexity for operators : log(n) 
    Extra Space Complexity for operators : o(1) .. No extra space
    Did this code successfully run on Leetcode : Yes
    Any problem you faced while coding this : No
    Initillay it will try go to the mid element. It target founds at mid then it will return.
    Then it will compare with the lower elemnt that means left most elment in the array.
    If nums[mid]>=nums[low] then check whether target is in between the range of mid and low.If it is then go to left side else on right side.
    Similarly for comparing mid with highest element.
    */

public class searchinRotatedSOrtedArray{


    public static void main(String args[]){
            int arr[] = new int[]{4,5,6,7,0,1,2};
            System.out.println(search(arr,0));
    }
    public static int search(int[] nums, int target) {
     
        int low =0;
        int high = nums.length-1;
        
        while(low<=high){
            int mid = low + (high-low)/2;
            if(nums[mid] == target)
                return  mid;
            else if(nums[low]<=nums[mid]){
                if(target>=nums[low] && target<nums[mid]){
                    high = mid-1;
                }else{
                    low = mid+1;
                }
            }else{
                if(target<=nums[high] && target>nums[mid]){
                    low = mid +1;
                }else{
                    high = mid -1;
                }
            }
        }
        return -1;
    }
}

File: searchinroratedsortedarray.py
"""
TC: O(logn)
Sc:O(1)
We will be checking which side is sorted and based upon that we will either re-assign value of low and high and recalculating the
value of mid untile mid==target, we will return the value of mid
:return
"""

class Solution:
    def searchRange(self, nums):
        if not nums or len(nums)==0:
            return -1
        low=0
        high=len(nums)-1
        while (low<=high):
            mid=(low+high)//2
            if nums[mid]==target:
                return mid
            elif nums[low]<=nums[mid]:
                print("left side is sorted")
                if nums[low]<=target and nums[mid]>target:
                    high=mid-1
                ##if elemeent is lying on the non sorted side
                else:
                    low=mid+1
            else:
                print("right side is sorted")
                if nums[mid]<target and nums[high]>=target:
                    low=mid+1
                ##if element is lying on the non sorted side
                else:
                    high=mid-1

        return 0
nums=[4,5,6,7,0,1,2]
target=5
solve=Solution()
print(solve.searchRange(nums))
